Harvesting to json

BevapDin opened the pull request:
This implement the money tree. Well, it _allows_ the money tree.

Add this to the money bundle item:

``` JSON
"seed_data": {
    "plant_name": "money tree",
    "fruit": "money_bundle",
    "byproducts": [ "log" ],
    "grow" : 1
},
```

Plant it, wait, harvest it. Gives you more money and a log (because it's a tree, get it?)

---

The seed data can be added to any item (tested with money bundle, which is generic item and the usual comestible seed).

"fruit" defines the item type of the fruit (count of it is determined by your survival skill, as before).
"byproducts" is a simple list of more items to spawn along (like withered plant, or straw, or ..., entries can be repeated - each entry is one item).

Note that the fruit item can be the same as the seed item (but that will give you more items as result because you get X fruits _and_ Y seed, X+Y items in total).

KA101 commented:
Tried this with an energy drink as the seed/fruit and an aluminum can as the byproduct.  Got a single can of energy drink back.  (I'd planted debug drink, not canned.)  Not sure if that's intended.

BevapDin commented:
Tested further:

Gave energy drink the seed data, wished for energy drink in a can. Planted it. Result is an empty can in inventory (good), similar a un-contained energy drink in the inventory is also used fine.

The planted item, however, is a can with energy drink as content (and cans are not seeds, so the game throws debug messages left and right). This is a bug.

I fixed this for now with the map editor: added an energy drink item, changed the furniture on top of it to harvestable plant. Harvesting works fine, result is (completely unskilled character) 2 cans with drink (one is the seed, the other is the fruit) and one empty can (byproduct). For a mega-skilled character (from survival level 24 on, you're guaranteed 12 fruits) it spawns 13 energy drinks (1 seed + 12 fruits) and one can.

So I changed it:
- Planting from within a container only adds the seed item, not the container.
- Harvested items don't spawn inside a container. Harvesting the energy drink spawns some energy drink on the ground.

----

While we're at at: should it be possible to spawn only fruits and no seeds at all (currently seeds are always spawned)?

facepalm commented:
>should it be possible to spawn only fruits and no seeds at all (currently seeds are always spawned)?

I hope so.  There's a number of plants whose fruits should be their own seeds - wheat, for example - as well as any plant whose fruit is actual fruit, with seeds inside, that technically ought to spawn when the fruit is eaten or processed in some manner.

Darkyhard commented:
Thyme, mugwort, wild herbs, dandelions and mushrooms shouldn't probably produce withered plants because you are using the whole plant in some way, not just its' seeds or leaves.

kevingranade commented:
Yes, producing fruit that may or may not itself contain seed should be a
feature, but not necessarily a blocker for this PR.  Ideally we could also
support a planting cycle like cutting -> bush -> cuttings and fruit.

> Harvested items don't spawn inside a container. Harvesting the energy
drink spawns some energy drink on the ground.

Of course not, you need to put a tap in an energy drink tree and attach a
bucket like with maple syrup harvesting :D (future feature request).
On Apr 16, 2015 7:36 AM, "Darkyhard" <notifications@github.com> wrote:

> Thyme, mugwort, wild herbs, dandelions and mushrooms shouldn't probably
> produce withered plants because you are using the whole plant in some way,
> not just its' seeds or leaves.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/12035#issuecomment-93750105>
> .
>

BevapDin commented:
>> should it be possible to spawn only fruits and no seeds at all (currently seeds are always spawned)?

> I hope so. There's a number of plants whose fruits should be their own seeds - wheat, for example - as well as any plant whose fruit is actual fruit, with seeds inside, that technically ought to spawn when the fruit is eaten or processed in some manner.

It is fixed. Tested and works fine for example with datura seeds:
```JSON
"seed_data": {
    "plant_name": "datura",
    "fruit": "datura_seed",
    "seeds": false,
    "grow" : 1
}
```
Harvesting only spawns some datura seeds. (Example not included in this PR.)

> Nice work, by the way. I hope the money tree stays in, it would be a cute easter egg.

Thanks. In that case, somebody (not me) has to add it. My PRs are far to serious business for such silly nonsense (-:

But as the money bundle has no further use in the game (except to burn it, but any generic plant product could be used for that) it might be fine.

> Thyme, mugwort, wild herbs, dandelions and mushrooms shouldn't probably produce withered plants because you are using the whole plant in some way, not just its' seeds or leaves.

I fixed that as well.

> Of course not, you need to put a tap in an energy drink tree and attach a
bucket like with maple syrup harvesting :D (future feature request).

Hmmm, a construction that uses a water faucet, some specific `construction::check` function, change plant furniture to some "plant with faucet" furniture with its own `iexamine` function. Add a syrup item on the tile to track the amount of extracted syrup, or maybe use the item variables of the seed item. I will keep that in mind.
